pest_derive/src/optimizer.rs

```diff
@@ -91,23 +91,6 @@ pub fn optimize(rules: Vec<Rule>) -> Vec<Rule> {
                     ty,
                     expr: expr.map_bottom_up(rotate_right)
                         .map_bottom_up(unroll_loops)
-                        .map_bottom_up(|expr| {
```

Comment:
I see the issue here: since strings are not owned anymore, concatenation is not possible. Seems like we'll have to revert this change.

Reply:
Do you happen to have any measurements comparing the string concatenation optimization with the slice optimization? The slice optimization prevents memory allocation, prevents copying and improves referential locality, whereas the string concatenation optimization reduces the number of nodes. On first glance I don't really see a clear cut advantage of one optimization over another. However, they can coexist :)
My proposal would be to keep the string slice references and to postpone the concatenation optimization for a separate PR. The reason for this is that the The concatenated strings could be represented later using another `Rule` variant containing a `String` (or another data structure, such as a `Vec[&'i str]`).

Reply:
The slice optimization happens to make compilation faster, while the concatenation works on improving parse performance once the actual code is generated which is why I would definitely go for the latter. Apart from that, the actual grammar parsing/optimization/generation step is very small during the whole compilation phase which is why I didn't bother at all to do any optimizations on this side and went for code clarity. (as best as I could while still trying to get features out 😃)

Having another variant for concatenations is not a bad idea, but wouldn't it flood the optimization and generation steps too much? Since this would require you to deal with multiple cases that, in the end, reach the same point. The reason why I find the `&str` to be advantageous is because it is way easier to write tests without all the ugly `.to_owned()` labels everywhere. However, this could be improved by writing some macros for test improvement. Quite frankly, tests are extremely verbose since I haven't had the time to invest in making them a bit nicer. However, I find their explicitness to be rather helpful. (if you kind of know your way around the code)

So, I would definitely go for the owned approach.